inc/templates.php

```diff
@@ -274,3 +279,212 @@ function convert_blocks_to_components( array $blocks ): array {
 
 	return $components;
 }
+
+/**
+ * Recursively iterate on a component tree.
+ *
+ * @param array $components Array of components.
+ * @return array
+ */
+function traverse_components( array $components ): array {
+
+	foreach ( $components as $index => &$component ) {
+
+		// This allows us to set text nodes.
+		if ( is_string( $component ) ) {
+			continue;
+		}
+
+		// Ensure we have all the right field and types.
+		$component = ensure_component_fields_exist( $component );
+
+		$component = handle_template_parts( $component );
+		$component = handle_data_provider( $component );
+		$component = handle_component_config_callbacks( $component );
+		$component = handle_component_callbacks( $component );
+
+		// Recursively loop though.
+		if ( ! empty( $component['children'] ) ) {
+			$component['children'] = traverse_components( $component['children'] );
+		}
+
+		// Ensure config and data providers are actually objects.
+		$component = validate_final_component( $component );
+	}
+
+	return $components;
+}
+
+
+/**
+ * Validate and typecast a component.
+ *
+ * @param array $component Component.
+ * @return array
+ */
+function ensure_component_fields_exist( $component ) {
+
+	if ( is_string( $component ) ) {
+		return $component;
+	}
+
+	return [
```

Comment:
This is fine, but would be a good spot to make use of `wp_parse_args()` for merging default config values with those present in the component.

Reply:
There's two reasons I did it this way,

1. So we only have the component fields, and other stuff gets filtered out (and I acknowledge there are a few ways to do this).
1. Typecasting